### DISCUSSION TITLE
Can we remove the `isnan` check in the `complex<T>` multiplication operator?

### DISCUSSION THREAD

**Discussion by jrhemstad** [NVIDIA/cccl#1031] [libcu++]
The `operator*(cuda::std::complex<T>, cuda::std::complex<T>)` does an [`isnan`](https://github.com/NVIDIA/libcudacxx/blob/aff156af14d94acf4afd5f8b780e8e6ef050d711/include/cuda/std/detail/libcxx/include/complex#L639) check on the resulting real and imaginary parts of the operation before returning the resulting `complex<T>`. 

Is this required by the C++ Standard?

At first glance, it would seem this is not required:

> If the result of a function is not mathematically defined or not in the range of representable values for its type, the behavior is undefined.

https://eel.is/c++draft/complex.numbers#general-3

If it is not required, can we remove these checks?

**Comment by jrhemstad** [comment]
@griwes @wmaxey @gonzalobg @brycelelbach 

Can any of you shine some light on this?

**Comment by jrhemstad** [comment]
CC @cliffburdick @luitjens

**Comment by jrhemstad** [comment]
This implementation appears to be inherited from the [C99 Standard Annex G](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf)

![latest-screenshot](https://user-images.githubusercontent.com/15221289/186777815-53690bfb-7be5-4e7b-9464-800f882e0f9a.png)

Given that the C++ Standard does not implicitly inherit everything from the C Standard and https://eel.is/c++draft/complex.numbers#general-3 makes no mention of any requirement about anything like this, I am convinced that we can remove this from our implementation and remain conforming to the C++ Standard.

**Comment by luitjens** [comment]
Ran a quick experiment where I modified 11.5 libcu++.  Modifications were to fix double cast and comment out isnan checks.   This data indicates the isnan checks are still very expensive on performance.  65us vs 110us.

The benchmark is found here:  https://github.com/NVIDIA/MatX/blob/main/bench/00_transform/conv.cu

A100 isnan checks, double fix
|                T                | Samples |  CPU Time  | Noise |  GPU Time  | Noise | Batch GPU  | Batch  |
|---------------------------------|---------|------------|-------|------------|-------|------------|--------|
|  cuda::std::__4::complex<float> |  31079x | 120.867 us | 1.40% | 112.339 us | 0.53% | 109.970 us | 31080x |

A100 NO isnan checks, double fix
|                T                | Samples |  CPU Time  | Noise |  GPU Time  | Noise | Batch GPU  | Batch  |
|---------------------------------|---------|------------|-------|------------|-------|------------|--------|
|  cuda::std::__4::complex<float> |  32944x |  76.357 us | 1.45% |  67.539 us | 1.59% |  64.615 us | 32945x |

**Comment by miscco** [comment]
I am wary of changing the current implementation because of non subtle runtime differences it would introduce. Those would only happen in a denormal case but still could potentially silently break user code.

Given that the chance of breaking something is rather slim and users might prefer the performance I would propose to add a feature macro that removes the sanitation of denormals.

Something in the direction `LIBCUDACXX_ENABLE_SIMPLIFIED_COMPLEX_OPERATIONS`?

Obviously I would like to tackle this in  #338

**Comment by miscco** [comment]
#338 because editing a comment does not link the issue / PR
